File: lib/services/backend_activation_service.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';
import 'package:fstapp/app_config.dart';
import 'package:fstapp/services/storage_helper.dart';
import 'package:http/http.dart' as http;

typedef BackendActivationRead = Future<String?> Function(String key);
typedef BackendActivationWrite = Future<void> Function(
    String key, String value);
typedef BackendActivationFetch = Future<List<int>> Function(Uri uri);

class ResolvedBackend {
  const ResolvedBackend({
    required this.supabaseUrl,
    required this.anonKey,
    required this.organizationId,
    required this.profileSha256,
    required this.isCanonical,
  });

  final String supabaseUrl;
  final String anonKey;
  final int organizationId;
  final String profileSha256;
  final bool isCanonical;

  String installationGeneration(String baseGeneration) => isCanonical
      ? '$baseGeneration/backend-canonical-v1/organization-$organizationId'
      : baseGeneration;
}

/// Resolves the one backend a process may use before Supabase is initialized.
///
/// Transition releases default to their compiled legacy backend. The only
/// remotely accepted instruction is the exact, build-approved canonical
/// activation document identified by its SHA-256. Once accepted, the local
/// marker is monotonic: an unavailable or older deployment can never route the
/// installation back to the legacy writer.
class BackendActivationService {
  BackendActivationService({
    String? tenantId,
    String? manifestUrl,
    String? canonicalManifestSha256,
    String? legacySupabaseUrl,
    String? legacyAnonKey,
    String? canonicalSupabaseUrl,
    String? canonicalAnonKey,
    int? legacyOrganizationId,
    int? canonicalOrganizationId,
    String? canonicalProfileSha256,
    BackendActivationRead? read,
    BackendActivationWrite? write,
    BackendActivationFetch? fetch,
  })  : tenantId = tenantId ?? AppConfig.backendActivationTenantId,
        manifestUrl = manifestUrl ?? AppConfig.backendActivationManifestUrl,
        canonicalManifestSha256 = canonicalManifestSha256 ??
            AppConfig.backendActivationCanonicalManifestSha256,
        legacySupabaseUrl = legacySupabaseUrl ?? AppConfig.effectiveSupabaseUrl,
        legacyAnonKey = legacyAnonKey ?? AppConfig.effectiveSupabaseAnonKey,
        canonicalSupabaseUrl = canonicalSupabaseUrl ??
            AppConfig.backendActivationCanonicalSupabaseUrl,
        canonicalAnonKey =
            canonicalAnonKey ?? AppConfig.backendActivationCanonicalAnonKey,
        legacyOrganizationId = legacyOrganizationId ?? AppConfig.organization,
        canonicalOrganizationId = canonicalOrganizationId ??
            AppConfig.backendActivationCanonicalOrganizationId,
        canonicalProfileSha256 = canonicalProfileSha256 ??
            AppConfig.backendActivationCanonicalProfileSha256,
        _read = read ?? ((key) => StorageHelper.get(key)),
        _write = write ?? ((key, value) => StorageHelper.set(key, value)),
        _fetch = fetch ?? _fetchManifest;

  static const int schemaVersion = 1;
  static const int canonicalGeneration = 1;
  static const String _markerPrefix = 'backend_activation_v1/';
  static const String _canonicalMarker = 'canonical:1';

  final String tenantId;
  final String manifestUrl;
  final String canonicalManifestSha256;
  final String legacySupabaseUrl;
  final String legacyAnonKey;
  final String canonicalSupabaseUrl;
  final String canonicalAnonKey;
  final int legacyOrganizationId;
  final int canonicalOrganizationId;
  final String canonicalProfileSha256;
  final BackendActivationRead _read;
  final BackendActivationWrite _write;
  final BackendActivationFetch _fetch;

  bool get isEnabled =>
      tenantId.isNotEmpty &&
      manifestUrl.isNotEmpty &&
      canonicalManifestSha256.isNotEmpty &&
      canonicalSupabaseUrl.isNotEmpty &&
      canonicalAnonKey.isNotEmpty &&
      canonicalOrganizationId > 0 &&
      RegExp(r'^[0-9a-f]{64}$').hasMatch(canonicalProfileSha256);

  Future<ResolvedBackend> resolve() async {
    if (!isEnabled) return _legacy;

    final markerKey = '$_markerPrefix$tenantId';
    try {
      if (await _read(markerKey) == _canonicalMarker) return _canonical;
    } catch (_) {
      return _legacy;
    }

    try {
      final bytes = await _fetch(Uri.parse(manifestUrl))
          .timeout(const Duration(seconds: 2));
      if (sha256.convert(bytes).toString() != canonicalManifestSha256) {
        return _legacy;
      }
      final value = jsonDecode(utf8.decode(bytes));
      if (value is! Map<String, dynamic> ||
          value.length != 4 ||
          value['schemaVersion'] != schemaVersion ||
          value['tenantId'] != tenantId ||
          value['generation'] != canonicalGeneration ||
          value['backend'] != 'canonical') {
        return _legacy;
      }
      await _write(markerKey, _canonicalMarker);
      return _canonical;
    } catch (_) {
      return _legacy;
    }
  }

  ResolvedBackend get _legacy => ResolvedBackend(
        supabaseUrl: legacySupabaseUrl,
        anonKey: legacyAnonKey,
        organizationId: legacyOrganizationId,
        profileSha256: '',
        isCanonical: false,
      );

  ResolvedBackend get _canonical => ResolvedBackend(
        supabaseUrl: canonicalSupabaseUrl,
        anonKey: canonicalAnonKey,
        organizationId: canonicalOrganizationId,
        profileSha256: canonicalProfileSha256,
        isCanonical: true,
      );

  static Future<List<int>> _fetchManifest(Uri uri) async {
    final response = await http.get(uri, headers: const {
      'cache-control': 'no-cache',
    });
    if (response.statusCode != 200) {
      throw StateError(
          'Backend activation manifest returned ${response.statusCode}');
    }
    return response.bodyBytes;
  }
}

File: web_client/src/services/backend_activation_service.js
import { AppConfig } from '../app_config.js';

const hex = (bytes) => [...new Uint8Array(bytes)]
    .map((value) => value.toString(16).padStart(2, '0'))
    .join('');

export class BackendActivationService {
    static schemaVersion = 1;
    static canonicalGeneration = 1;
    static markerPrefix = 'festapp-backend-activation-v1:';
    static canonicalMarker = 'canonical:1';

    constructor({ config = AppConfig, storage,
        fetchImpl = globalThis.fetch?.bind(globalThis), subtle = globalThis.crypto?.subtle } = {}) {
        this.config = config;
        try {
            this.storage = storage === undefined ? globalThis.localStorage : storage;
        } catch (_) {
            this.storage = null;
        }
        this.fetchImpl = fetchImpl;
        this.subtle = subtle;
    }

    get enabled() {
        return Boolean(
            this.config.backendActivationTenantId &&
            this.config.backendActivationManifestUrl &&
            this.config.backendActivationCanonicalManifestSha256 &&
            this.config.backendActivationCanonicalSupabaseUrl &&
            this.config.backendActivationCanonicalAnonKey
        );
    }

    async resolve() {
        if (!this.enabled) return this.legacy;
        const markerKey = BackendActivationService.markerPrefix +
            this.config.backendActivationTenantId;
        if (!this.storage) return this.legacy;
        try {
            if (this.storage.getItem(markerKey) === BackendActivationService.canonicalMarker) {
                return this.canonical;
            }
        } catch (_) {
            return this.legacy;
        }
        if (!this.fetchImpl || !this.subtle) return this.legacy;

        let timeoutId;
        try {
            const response = await Promise.race([
                this.fetchImpl(
                    `${this.config.backendActivationManifestUrl}?activation=${Date.now()}`,
                    { cache: 'no-store', headers: { 'cache-control': 'no-cache' } },
                ),
                new Promise((_, reject) => {
                    timeoutId = setTimeout(
                        () => reject(new Error('Backend activation lookup timed out')), 2000,
                    );
                }),
            ]);
            if (!response.ok) return this.legacy;
            const bytes = await response.arrayBuffer();
            const digest = hex(await this.subtle.digest('SHA-256', bytes));
            if (digest !== this.config.backendActivationCanonicalManifestSha256) {
                return this.legacy;
            }
            const value = JSON.parse(new TextDecoder().decode(bytes));
            if (Object.keys(value).length !== 4 ||
                value.schemaVersion !== BackendActivationService.schemaVersion ||
                value.tenantId !== this.config.backendActivationTenantId ||
                value.generation !== BackendActivationService.canonicalGeneration ||
                value.backend !== 'canonical') {
                return this.legacy;
            }
            this.storage.setItem(markerKey, BackendActivationService.canonicalMarker);
            if (this.storage.getItem(markerKey) !== BackendActivationService.canonicalMarker) {
                return this.legacy;
            }
            return this.canonical;
        } catch (_) {
            return this.legacy;
        } finally {
            clearTimeout(timeoutId);
        }
    }

    get legacy() {
        return Object.freeze({
            supabaseUrl: this.config.supabaseUrl,
            anonKey: this.config.anonKey,
            isCanonical: false,
        });
    }

    get canonical() {
        return Object.freeze({
            supabaseUrl: this.config.backendActivationCanonicalSupabaseUrl,
            anonKey: this.config.backendActivationCanonicalAnonKey,
            isCanonical: true,
        });
    }
}
